text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,170 @@
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    return (ctx, next) => {
+      const {serialize, deserialize} = SharedState(ctx, 'some-id');
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize({
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+
+// in tests
+import {SharedStateToken, MockSharedState} from 'fusion-core';
+app.register(SharedStateToken, MockSharedState({
+  'some-id': {
+    some: 'json'
+  },
+  'other-data': {}
+}));
+```
+
+# Motivation
+
+It is a very common pattern in fusion to have some shared state between the server and
+the browser encapsulated into a plugin. The current model for this has users referencing
+`ctx.template` directly on the server, and querying the DOM directly from the browser.
+
+This can make testing difficult, especially when running browser tests since there is not
+an easy way to mock the deserialization other than each plugin exposing their own deserialization
+token. 
+
+Providing a standard API for sharing json data between server and browser plugins
+will make it easier to test these types of plugins in isolation without needing to resort 
+to integration tests. Additionally, it can help solve problems where users may want custom 
+serialization / deserialization mechanisms, such as using tools like immutable.js.
+
+# Detailed design
+
+This proposal adds the following APIs to `fusion-core` 
+
+### `SharedStateToken`
+
+```js
+import {SharedStateToken} from 'fusion-core';
+```
+
+The `SharedStateToken` will have a default implementation registered onto each fusion app
+which provides a `{serialize}` and `{deserialize}` function which serialize json data on the
+server into a `<script type='application/json'>` dom element which is appended to `ctx.template.body`, and
+deserialize the data from the dom node in the browser. Example Usage:
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    return (ctx, next) => {
+      const {serialize, deserialize} = SharedState(ctx, 'some-id');
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize({
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+```
+
+### `MockSharedState`
+
+```js
+import {MockSharedState} from 'fusion-core';
+```
+
+This API will provide an easy way to provide mock data for all deserialize
+calls in an application. 
+
+```js
+import {MockSharedState, SharedStateToken} from 'fusion-core';
+// Deserialize calls will now return the mock values for these ids
```

Comment:
I think a couple aspects of this particular explanation aren't totally clear to me
- What's the difference between setting the shared state, setting the default shared state, and mocking the shared state?
- Does this change which "deserialize" function is called, or merely change the values on which deserialize operates on? Part of this explanation seems to imply that `deserialize` is mocked, but the name of the actual function makes it seem like only the itself data is being changed.

Reply:
Yea open for suggestions on the naming here. I actually originally put the `MockSharedState` (or w/e we call it) into `fusion-test-utils` because it really should only be used for testing. It is really just a helper for mocking `deserialize`. Maybe it could mock `serialize` also, but I think that will be less useful. But that comment I think describes it pretty well:

> Deserialize calls will now return the mock values for these ids

It does just that.